pkg/clusters/addons/loadimage/addon.go

```diff
@@ -0,0 +1,65 @@
+package loadimage
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/kong/kubernetes-testing-framework/pkg/clusters"
+	"github.com/kong/kubernetes-testing-framework/pkg/clusters/types/kind"
+)
+
+// -----------------------------------------------------------------------------
+// CertManager Addon
+// -----------------------------------------------------------------------------
+
+const (
+	// AddonName indicates the unique name of this addon.
+	AddonName clusters.AddonName = "loadimage"
+)
+
+type Addon struct {
+	image  string
+	loaded bool
+}
+
+func New() clusters.Addon {
+	return &Addon{}
+}
+
+// -----------------------------------------------------------------------------
+// CertManager Addon - Addon Implementation
+// -----------------------------------------------------------------------------
+
+func (a *Addon) Name() clusters.AddonName {
+	return AddonName
+}
+
+func (a *Addon) Deploy(ctx context.Context, cluster clusters.Cluster) error {
+	if a.image == "" {
+		a.loaded = true
+		return nil
+	}
+
+	switch ctype := cluster.Type(); ctype {
+	case kind.KindClusterType:
+		if err := a.loadIntoKind(cluster); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("loadimage addon is not supported by cluster type '%v'", cluster.Type())
+	}
+
+	return nil
+}
+
+func (a *Addon) Delete(ctx context.Context, cluster clusters.Cluster) error {
+	// per https://github.com/kubernetes-sigs/kind/issues/658 this is basically impossible
+	return fmt.Errorf("cannot remove loaded images from a cluster")
```

Comment:
"what addons are today" very much needs a definition to be a useful bar if it's something beyond satisfying the interface, which appears to be the implication here 🙂 My intuitive definition was "almost anything we add to the test cluster after it starts"; the implied definition appears to instead be "things you can add by applying a Kubernetes manifest, nothing else", or maybe a slightly broader "things that function the same way on all Kubernetes implementations". If it's one of the latter two, IMO KTF needs some new, similar interface for things that aren't manifests, but also aren't something we'd add to the Cluster interface instead, without passing the buck for dealing with all implementation idiosyncrasies downstream to test clients.

We can't shield them from _all_ idiosyncrasies, but we can probably reasonably shield them from most. In this specific case, you shouldn't much care about deleting Docker images; you can just upload a new one with the same name to replace it. Some domain-specific knowledge of what the addon is doing will probably be necessary unless you really constrain what addons are.

I considered this fine because I wasn't aware of any scenarios where you'd call an addon's `Delete()` without knowing the specific addon, and if you did try to delete this one during the course of development, you'd learn it's not possible on the first attempt and not try it afterwards (or be so determined to purge the image that you contribute `kind delete docker-image` upstream).

Currently we don't call the `Delete()`s anywhere, so there's no practical use cases to draw from. This would fail if you did a blind `ListAddons()` and then `DeleteAddon()` on all results that aborted on any failure. That's only viable if we say "don't do that" (honestly, if you want that delete the cluster and make a new one; eat the extra time) or create the suggested mechanism that reports a sort of benign non-error failure.

It would be possible in other implementations, e.g. [GCR does let you delete images](https://cloud.google.com/container-registry/docs/managing#deleting_images), but that doesn't remove the issue that some cluster types will always fail.

25f0846 splits the cases and makes KIND's delete a noop.